onnxruntime/core/providers/nnapi/nnapi_builtin/builders/shaper.cc

```diff
@@ -371,25 +371,29 @@ void Shaper::AddShape(const std::string& name, const Shape& shape) {
   shape_map_[name] = shape;
 }
 
-void Shaper::UpdateShape(const std::string& name, const Shape& new_shape) {
-  ORT_ENFORCE(shaper_finalized_,
-              "Cannot UpdateShape while shaper is not finalized");
+Status Shaper::UpdateShape(const std::string& name, const Shape& new_shape) {
+  ORT_RETURN_IF_NOT(shaper_finalized_,
+                    "Cannot UpdateShape while shaper is not finalized");
 
   const Shape& old_shape = shape_map_.at(name);
   if (old_shape != new_shape) {
-    if (Product(old_shape) != 0)
-      ORT_THROW("The shape should be same size or old shape has size 0 (dynamic shape)");
+    ORT_RETURN_IF_NOT(Product(old_shape) == 0,
+                      "The shape should be same size or old shape has size 0 (dynamic shape)");
 
```

Comment:
Do we need to implicitly use a shape of size 0 to indicate a dynamic shape? Given we have a Shape class could that explicitly track whether it's dynamic instead?

Reply:
This is at execution time, at this time, we do not have access to the ONNX shape anymore since the execution is entirely in NNAPI


---
In reply to: [462104113](https://github.com/microsoft/onnxruntime/pull/4650#discussion_r462104113) [](ancestors = 462104113)